### PR TITLE
feat(derive): BatchQueue Update [Holocene]

### LINKED ISSUE #560: [HOLOCENE] feat(derive): Batch Queue Changes
### Description

> [!WARNING]
>
> It is highly encouraged to [read the specs](https://specs.optimism.io/protocol/holocene/overview.html) prior to working on this implementation.

Holocene's strict ordering changes the behavior of batches within the derivation pipeline. Since batches now arrive in strict order, unordered batches [can now be immediately dropped](https://specs.optimism.io/protocol/holocene/derivation.html#batch-queue).
If a batch is invalid and dropped, it's span batch is also discarded.

## Output

- [ ] Modify the `BatchQueue` in a **backwards-compatible** way such that batches that are found to have a `BatchValidity::Future` are now dropped.
  - In the [`BatchQueue::derive_next_batch`](https://github.com/anton-rs/kona/blob/main/crates/derive/src/stages/batch_queue.rs#L136-L166) method, only add a batch with `BatchValidity::Future` to the remaining list pre-holocene. This can be checked using [`RollupConfig::is_holocene_active`](https://github.com/alloy-rs/op-alloy/blob/main/crates/genesis/src/rollup.rs#L207-L210).
 	```rust
 	// -- snip --

 	BatchValidity::Future => {
	  if !self.cfg.is_holocene_active(timestamp) {
	    remaining.push(batch.clone());
	  }
    }

    // -- snip --
    ```

### PR DESCRIPTION
**Pull request by refcell**
### Description

Updates the `BatchQueue` for Holocene.

### Metadata

Closes #560

### CODE REVIEW

**Comment by refcell** [comment]
* **#601** <a href="https://app.graphite.dev/github/pr/anton-rs/kona/601?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @refcell and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by refcell** [comment]
Need to add more tests to the `BatchQueue`...

File: crates/derive/src/stages/batch_queue.rs [hunk 1]
```diff
@@ -151,7 +153,14 @@ where
...
             match validity {
                 BatchValidity::Future => {
-                    remaining.push(batch.clone());
+                    // Drop Future batches post-holocene.
+                    //
+                    // See: <https://specs.optimism.io/protocol/holocene/derivation.html#batch-queue>
+                    if !self.cfg.is_holocene_active(origin.timestamp) {
+                        remaining.push(batch.clone());
+                    } else {
+                        warn!(target: "batch-queue", "[HOLOCENE] Dropping future batch with parent: {}", parent.block_info.number);
```
Comment (clabby) [inline crates/derive/src/stages/batch_queue.rs:162]: We should also flush the prev stage if we drop something here, yes?

File: crates/derive/src/stages/batch_queue.rs [hunk 2]
```diff
@@ -243,10 +251,17 @@ where
...
         // If we drop the batch, validation logs the drop reason with WARN level.
-        if data.check_batch(&self.cfg, &self.l1_blocks, parent, &mut self.fetcher).await.is_drop() {
+        let validity =
+            data.check_batch(&self.cfg, &self.l1_blocks, parent, &mut self.fetcher).await;
+        if validity.is_drop() {
             self.prev.flush();
             return Ok(());
         }
+        // Post-Holocene, future batches are dropped due to prevent gaps.
+        if self.cfg.is_holocene_active(origin.timestamp) && validity.is_future() {
```
Comment (clabby) [inline crates/derive/src/stages/batch_queue.rs:261]: same same - prob best to just combine this check i.e. `validity.is_drop() || (self.cfg.is_holocene_active(origin.timestamp) && validity.is_future())`

File: crates/derive/src/stages/batch_queue.rs (hunk 1 above)
Comment (refcell) [reply crates/derive/src/stages/batch_queue.rs:162]: I don't think so.

File: crates/derive/src/stages/batch_queue.rs [hunk 3]
```diff
@@ -243,7 +252,12 @@ where
         let origin = self.origin.ok_or(PipelineError::MissingOrigin.crit())?;
         let data = BatchWithInclusionBlock { inclusion_block: origin, batch };
         // If we drop the batch, validation logs the drop reason with WARN level.
-        if data.check_batch(&self.cfg, &self.l1_blocks, parent, &mut self.fetcher).await.is_drop() {
+        let validity =
+            data.check_batch(&self.cfg, &self.l1_blocks, parent, &mut self.fetcher).await;
+        // Post-Holocene, future batches are dropped due to prevent gaps.
+        let drop = validity.is_drop() ||
+            self.cfg.is_holocene_active(origin.timestamp) && validity.is_future();
```
Comment (clabby) [inline crates/derive/src/stages/batch_queue.rs:259]:
```suggestion
            (self.cfg.is_holocene_active(origin.timestamp) && validity.is_future());
```

**Review by clabby — APPROVED** [review]

**Pull request merged**